File: src/mcp/tools/definitions/subplot_create.ts
/**
 * subplot_createツール定義
 * サブプロット要素を作成するMCPツール
 */

import type {
  McpToolDefinition,
  ToolExecutionContext,
} from "@storyteller/mcp/tools/tool_registry.ts";
import { executeCliCommand } from "@storyteller/mcp/tools/cli_adapter.ts";
import { ElementSubplotCommand } from "@storyteller/cli/modules/element/subplot.ts";

export const subplotCreateTool: McpToolDefinition = {
  name: "subplot_create",
  description:
    "サブプロット要素を作成します。物語のプロット構造（メインプロット、サブプロット、並行プロット、背景プロット）を管理するための構造を定義します。",
  inputSchema: {
    type: "object",
    properties: {
      name: {
        type: "string",
        description: "サブプロット名",
      },
      type: {
        type: "string",
        enum: ["main", "subplot", "parallel", "background"],
        description:
          "プロットタイプ: main（メインプロット）, subplot（サブプロット）, parallel（並行プロット）, background（背景プロット）",
      },
      summary: {
        type: "string",
        description: "サブプロットの概要",
      },
      importance: {
        type: "string",
        enum: ["major", "minor", "subtle"],
        description: "重要度: major（主要）, minor（副次）, subtle（微妙）",
      },
      // Why: McpPropertySchemaがネストされたobject propertiesをサポートしないため、
      // "characterId:weight" 形式の文字列配列で表現する
      focusCharacters: {
        type: "array",
        items: { type: "string" },
        description:
          "フォーカスキャラクター（\"characterId:weight\"形式の配列。weight: primary/secondary）",
      },
      parentSubplot: {
        type: "string",
        description: "親サブプロットID",
      },
    },
    required: ["name", "type", "summary"],
  },
  execute: async (
    args: Record<string, unknown>,
    context?: ToolExecutionContext,
  ) => {
    const name = args.name as string | undefined;
    const type = args.type as string | undefined;
    const summary = args.summary as string | undefined;

    if (!name || typeof name !== "string" || name.trim() === "") {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'name' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    if (!type || typeof type !== "string") {
      return {
        content: [
          {
            type: "text" as const,
            text:
              "Error: 'type' parameter is required (main|subplot|parallel|background).",
          },
        ],
        isError: true,
      };
    }

    const validTypes = ["main", "subplot", "parallel", "background"];
    if (!validTypes.includes(type)) {
      return {
        content: [
          {
            type: "text" as const,
            text: `Error: Invalid type '${type}'. Must be one of: ${
              validTypes.join(", ")
            }.`,
          },
        ],
        isError: true,
      };
    }

    if (!summary || typeof summary !== "string" || summary.trim() === "") {
      return {
        content: [
          {
            type: "text" as const,
            text: "Error: 'summary' parameter is required.",
          },
        ],
        isError: true,
      };
    }

    // CLIコマンド用の引数を構築
    const commandArgs: Record<string, unknown> = {
      name,
      type,
      summary,
    };

    if (typeof args.importance === "string") {
      commandArgs.importance = args.importance;
    }

    if (Array.isArray(args.focusCharacters)) {
      const chars = (args.focusCharacters as unknown[])
        .filter((c) => typeof c === "string")
        .map((c) => (c as string).trim())
        .filter((c) => c.length > 0);
      if (chars.length > 0) {
        commandArgs["focus-characters"] = chars.join(",");
      }
    }

    if (typeof args.parentSubplot === "string") {
      commandArgs["parent-subplot"] = args.parentSubplot;
    }

    const handler = new ElementSubplotCommand();
    return await executeCliCommand(handler, commandArgs, context?.projectRoot);
  },
};

File: internal/mcp/tools/subplot_create.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"

	"github.com/takets/street-storyteller/internal/domain"
	"github.com/takets/street-storyteller/internal/mcp/protocol"
)

// subplotSlugRe strips characters unsuitable for a slug ID.
var subplotSlugRe = regexp.MustCompile(`[^a-z0-9]+`)

// subplotSlugify converts a human-readable name into a lowercase underscore id.
// Why: local helper to avoid depending on a package-level slugify that may not
// exist in this tools package; mirrors the convention in timeline_create.go.
func subplotSlugify(name string) string {
	s := strings.ToLower(name)
	s = subplotSlugRe.ReplaceAllString(s, "_")
	s = strings.Trim(s, "_")
	if s == "" {
		return fmt.Sprintf("sp_%x", name)
	}
	return s
}

// SubplotCreateTool validates inputs and returns the generated Subplot stub.
// Why: persistence is out of scope — consistent with the pattern used by
// timeline_create and foreshadowing_create; the tool only validates and
// serialises the domain struct so the caller can inspect inputs before writing.
type SubplotCreateTool struct{}

type subplotCreateArgs struct {
	Name    string `json:"name"`
	Type    string `json:"type"`
	Summary string `json:"summary"`
	ID      string `json:"id"`
}

var validSubplotTypes = map[string]domain.SubplotType{
	"main":       domain.SubplotTypeMain,
	"subplot":    domain.SubplotTypeSubplot,
	"parallel":   domain.SubplotTypeParallel,
	"background": domain.SubplotTypeBackground,
}

// Definition advertises the subplot_create schema.
func (SubplotCreateTool) Definition() protocol.Tool {
	return protocol.Tool{
		Name:        "subplot_create",
		Description: "Create a new subplot and return its structure as JSON",
		InputSchema: json.RawMessage(`{
			"type":"object",
			"required":["name","type","summary"],
			"properties":{
				"name":{"type":"string","description":"Human-readable subplot name"},
				"type":{"type":"string","enum":["main","subplot","parallel","background"],"description":"Narrative role of the subplot"},
				"summary":{"type":"string","description":"Brief summary of the subplot"},
				"id":{"type":"string","description":"Optional ID; derived from name if omitted"}
			}
		}`),
	}
}

// Handle validates args and returns the generated Subplot stub as text.
func (SubplotCreateTool) Handle(_ context.Context, args json.RawMessage, _ ExecutionContext) (*protocol.CallToolResult, error) {
	var a subplotCreateArgs
	if len(args) > 0 {
		_ = json.Unmarshal(args, &a)
	}

	if a.Name == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "name is required"}},
			IsError: true,
		}, nil
	}
	if a.Type == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "type is required"}},
			IsError: true,
		}, nil
	}
	if a.Summary == "" {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: "summary is required"}},
			IsError: true,
		}, nil
	}

	st, ok := validSubplotTypes[a.Type]
	if !ok {
		return &protocol.CallToolResult{
			Content: []protocol.ContentBlock{{Type: "text", Text: fmt.Sprintf("invalid type %q: must be one of main, subplot, parallel, background", a.Type)}},
			IsError: true,
		}, nil
	}

	id := a.ID
	if id == "" {
		id = subplotSlugify(a.Name)
	}

	sp := domain.Subplot{
		ID:      id,
		Name:    a.Name,
		Type:    st,
		Status:  domain.SubplotStatusActive,
		Summary: a.Summary,
		Beats:   []domain.PlotBeat{},
	}

	b, _ := json.Marshal(sp)
	return &protocol.CallToolResult{
		Content: []protocol.ContentBlock{{
			Type: "text",
			Text: fmt.Sprintf("subplot created: %s\n%s", id, string(b)),
		}},
	}, nil
}
